### PR TITLE
Implement CPU plugin just-in-time emitter for IsFinite operation

### PR DESCRIPTION
**Pull request by awayzjj**
Closes https://github.com/openvinotoolkit/openvino/issues/24416
### Details:
 - *item1*
 - *...*

### Tickets:
 - [CVS-137697](https://jira.devtools.intel.com/browse/CVS-137697)

### CODE REVIEW

**Comment by awayzjj** [comment]
@eshoguli @dmitry-gorokhov  Hi, I have implemented this new emitter. Could you please help me review the PR and trigger the CI? Thank you very much!

**Comment by awayzjj** [comment]
Hello @dmitry-gorokhov @eshoguli , just a gentle reminder to review this PR when you have time.

**Comment by awayzjj** [comment]
@mlukasze Hi, could you please help me trigger the CI?

**Comment by mlukasze** [comment]
build_jenkins

**Comment by awayzjj** [comment]
Hello @dmitry-gorokhov  @eshoguli @mlukasze Some tests failed :(

The visible failed test is about JAX FE which seems not related with my PR ?
Could you please help me check other failed tests?

**Comment by awayzjj** [comment]
@dmitry-gorokhov  @eshoguli Gentle ping.

**Comment by awayzjj** [comment]
@dmitry-gorokhov @eshoguli Just a friendly reminder for help when you have time. I want to finish this emitter and move to another!

**Comment by a-sidorova** [comment]
@awayzjj there are two failed jobs which are not affected by your changes.

I see that at the moment the both of them are green on the master branch, so I suggest to just rebase your branch.

**Comment by awayzjj** [comment]
@a-sidorova Thank you so much! I have rebased my branch, Could you please help me trigger the CI?

**Comment by mlukasze** [comment]
build_jenkins

**Comment by awayzjj** [comment]
@a-sidorova Hi, there are 2 jobs that have been hanging for a long time, and 3 jobs have failed. Could you please provide some suggestions on how to address these issues? Thank you very much!

**Comment by a-sidorova** [comment]
@awayzjj hi! To fix GPU jobs, you can rebase to the latest master to get the commit https://github.com/openvinotoolkit/openvino/pull/25355 which fixes the failed test. As for 2 hanging GHA jobs, there was a queue.

**Comment by awayzjj** [comment]
@a-sidorova Thank you so much! I have rebased my branch, Could you please help me trigger the CI again?

**Comment by a-sidorova** [comment]
build_jenkins

**Comment by awayzjj** [comment]
@a-sidorova @mlukasze Wow, the CI is full green. Can you merge the PR? Thank you very much!

**Review by eshoguli — COMMENTED** [review]
New emitter handles special values: `Nan` and `Infinity`, but input tensor generation was not changed. Please, update `ActivationLayerCPUTest::generate_inputs` to have `Nan` and `Infinity` values in test input tensor. Note, please, there is an example for `Sign` activation here. You can do something similar but use more then one value, please.

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 1]
```diff
@@ -802,6 +802,79 @@ std::set<std::vector<element::Type>> jit_hswish_emitter::get_supported_precision
...
+    using TReg = typename dnnl::impl::cpu::aarch64::cpu_isa_traits<isa>::TReg;
+
+    TReg src = TReg(in_vec_idxs[0]);
+    TReg dst = TReg(out_vec_idxs[0]);
+    TReg aux = TReg(aux_vec_idxs[0]);
+
+    // According to the IEEE standard, NaN values have the odd property that comparisons involving them are always false.
+    h->fcmeq(dst.s, src.s, src.s);
+    h->ld1r(aux.s, table_val2("zero"));
+    h->fcmeq(dst.s, dst.s, aux.s);
```
Comment (eshoguli) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:856]: Replace: 
```
    h->ld1r(aux.s, table_val2("zero"));
    h->fcmeq(dst.s, dst.s, aux.s);
```
with one line, please: 
```
    h->not_(dst.b16, dst.b16);
```

Note, you don't need `zero`.

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp [hunk 2]
```diff
@@ -802,6 +802,79 @@ std::set<std::vector<element::Type>> jit_hswish_emitter::get_supported_precision
...
+
+    h->fabs(src.s, src.s);
+    h->ld1r(aux.s, table_val2("inf"));
+    h->fcmeq(src.s, src.s, aux.s);
+
+    h->orr(dst.b16, dst.b16, src.b16);
+
+    h->ld1r(aux.s, table_val2("zero"));
+    h->fcmeq(dst.s, dst.s, aux.s);
```
Comment (eshoguli) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:865]: Replace with one line, please: `h->not_(dst.b16, dst.b16);`. You don't need `zero`.

File: src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp (hunk 1 above)
Comment (eshoguli) [inline src/plugins/intel_cpu/src/emitters/plugin/aarch64/jit_eltwise_emitters.cpp:856]: `dst` register can be the same as `src` register. Use auxiliary register as source in `fcmeq`.

**Comment by awayzjj** [comment]
@eshoguli Hi, thank you very much for your review! I have incorporated your suggestions, could you please review the PR again?

**Comment by mlukasze** [comment]
build_jenkins

File: src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp [hunk 3]
```diff
@@ -95,6 +95,15 @@ void ActivationLayerCPUTest::generate_inputs(const std::vector<ov::Shape>& targe
...
             if ((activationType == utils::ActivationTypes::Sign) && funcInput.get_element_type() == ov::element::f32) {
                 static_cast<float*>(tensor.data())[0] = std::numeric_limits<float>::quiet_NaN();
+            } else if ((activationType == utils::ActivationTypes::IsFinite) && funcInput.get_element_type() == ov::element::f32 && tensor.get_size() >= 8) {
+                static_cast<float*>(tensor.data())[0] = std::numeric_limits<float>::quiet_NaN(); // nan
+                static_cast<float*>(tensor.data())[1] = std::numeric_limits<float>::signaling_NaN(); // nan
+                static_cast<float*>(tensor.data())[2] = std::acos(2); // nan
+                static_cast<float*>(tensor.data())[3] = std::tgamma(-1); // nan
+                static_cast<float*>(tensor.data())[4] = std::log(-1); // nan
+                static_cast<float*>(tensor.data())[5] = std::sqrt(-1); // -nan
+                static_cast<float*>(tensor.data())[6] = 0 / 0.0; // -nan
```
Comment (dmitry-gorokhov) [inline src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp:105]: @awayzjj This line leads to failed Win build on CI: 
"openvino/src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp(105): error C2124: divide or mod by zero"

Comment (awayzjj) [reply src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp:105]: @dmitry-gorokhov Thank you! But I do not have much experirence on windows. Should I just remove the line 105?

Comment (dmitry-gorokhov) [reply src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp:105]: You are implementing isFinite operation. So I believe it is enough to have 4 specific values:  nan, -nan, inf, -inf

Comment (awayzjj) [reply src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp:105]: Thank you, I will fix it later today!

**Comment by dmitry-gorokhov** [comment]
build_jenkins

File: src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp [hunk 4]
```diff
@@ -95,6 +95,12 @@ void ActivationLayerCPUTest::generate_inputs(const std::vector<ov::Shape>& targe
             // cover Sign NAN test case
             if ((activationType == utils::ActivationTypes::Sign) && funcInput.get_element_type() == ov::element::f32) {
                 static_cast<float*>(tensor.data())[0] = std::numeric_limits<float>::quiet_NaN();
+            } else if ((activationType == utils::ActivationTypes::IsFinite) && funcInput.get_element_type() == ov::element::f32 && tensor.get_size() >= 5) {
+                static_cast<float*>(tensor.data())[0] = std::numeric_limits<float>::quiet_NaN(); // nan
```
[the remark this answers is not in the document]
Comment (awayzjj) [reply src/plugins/intel_cpu/tests/functional/custom/single_layer_tests/classes/activation.cpp:99]: @eshoguli I have already added the check(`tensor.get_size() >= 5`) at the end of the if statement , is it ok?

**Comment by dmitry-gorokhov** [comment]
@awayzjj Could you please rebase you PR on latest master state? This is required to make CI green.

**Comment by mg-intel** [comment]
build_jenkins

**Comment by awayzjj** [comment]
@dmitry-gorokhov @eshoguli Hi, the CI is still not full green after rebasing :(

**Review by dmitry-gorokhov — APPROVED** [review]

**Comment by dmitry-gorokhov** [comment]
@awayzjj CI is green now. Added in merge queue. 
Great job!

**Comment by awayzjj** [comment]
@dmitry-gorokhov Thank you so much for your patience and help! 

I have only one emitter to implement! Before starting, I have one question: how much performance improvements in terms of speed and throughput can we expect from JIT? Are there approximate numbers you can share?

**Comment by dmitry-gorokhov** [comment]
@awayzjj 
> I have only one emitter to implement!

We have two more that need contributors. Feel free to take :)
image

As for performance improvements it depends on the particular operation. As you know except JIT we have 2 backends ACL and reference impls. 
JIT usually delivers 10-50% better perf than ACL. Comparing with reference the difference is much more noticable (since reference impl is very slow itself): 10-100x faster.

**Comment by awayzjj** [comment]
@dmitry-gorokhov  I've taken these two issues and will work on resolving them as quickly as possible. I hope that in the future, you can release slightly different issues, as I want to become more familiar with our ARM backend! :)

**Pull request merged**